File: SubscriptionApi/SubscriptionApi/ResponseBuilders/SubscriptionResponseBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SubscriptionApi.Dto;

namespace SubscriptionApi.ResponseBuilders
{
  public class SubscriptionResponseBuilder : 
    SubscriptionStopResponseBuilder, SubscriptionStartResponseBuilder
  {
    private readonly List<string> _errors = new List<string>();

    public void AssertNoFatalErrors(Exception exceptionToThrow)
    {
      if (_errors.Any())
      {
        throw exceptionToThrow;
      }
    }

    public void NotAuthorized(string assetName, string userName)
    {
      _errors.Add(userName + " has no rights to " + assetName);
    }

    public void NotValid(string dataName)
    {
      _errors.Add(dataName + " is invalid");
    }

    StartSubscriptionResponseDto SubscriptionStartResponseBuilder.Build()
    {
      return new StartSubscriptionResponseDto()
      {
        Failure = _errors.Any(),
        Errors = _errors.ToArray()
      };
    }
    public StopSubscriptionResponseDto Build()
    {
      return new StopSubscriptionResponseDto()
      {
        Failure = _errors.Any(),
        Errors = _errors.ToArray()
      };
    }

    public void NoResolutionResultsFor(string name)
    {
      _errors.Add("Resolution of " + name + " yielded no results");
    }

    public void NoSubscriptionToStopWithId(string subscriptionId)
    {
      _errors.Add("No subscription with id " + subscriptionId + " found");
    }

    public void UserNotAuthorized(string userName)
    {
      _errors.Add("User " + userName + " could not be authorized in the system");
    }

    public void NotAuthorizedForAsset(string assetName, string userName)
    {
      _errors.Add("User " + userName + " is not authorized for " + assetName);
    }
  }

}

File: Java/factories/_01_no_factories/src/main/java/ResponseBuilders/SubscriptionResponseBuilder.java
package responseBuilders;

import dto.StartSubscriptionResponseDto;
import dto.StopSubscriptionResponseDto;

import java.util.ArrayList;
import java.util.List;

public class SubscriptionResponseBuilder implements
    SubscriptionStopResponseBuilder, SubscriptionStartResponseBuilder {
    private final List<String> errors = new ArrayList<>();

    public void assertNoFatalErrors(RuntimeException exceptionToThrow) {
        if (!errors.isEmpty()) {
            throw exceptionToThrow;
        }
    }

    public void notAuthorized(String assetName, String userName) {
        errors.add(userName + " has no rights to " + assetName);
    }

    public void notValid(String dataName) {
        errors.add(dataName + " is invalid");
    }

    public StartSubscriptionResponseDto buildStart() //todo rename to build start
    {
        StartSubscriptionResponseDto dto = new StartSubscriptionResponseDto();
        dto.failure = !errors.isEmpty();
        dto.errors = errors.stream().toArray(String[]::new);
        return dto;
    }

    public StopSubscriptionResponseDto buildStop() {
        StopSubscriptionResponseDto dto = new StopSubscriptionResponseDto();
        dto.failure = !errors.isEmpty();
        dto.errors = errors.stream().toArray(String[]::new);
        return dto;
    }

    public void noResolutionResultsFor(String name) {
        errors.add("Resolution of " + name + " yielded no results");
    }

    public void noSubscriptionToStopWithId(String subscriptionId) {
        errors.add("No subscription with id " + subscriptionId + " found");
    }

    public void userNotAuthorized(String userName) {
        errors.add("User " + userName + " could not be authorized in the system");
    }

    public void notAuthorizedForAsset(String assetName, String userName) {
        errors.add("User " + userName + " is not authorized for " + assetName);
    }
}

